### ISSUE TITLE
geoPath keeping track of streamed features to a canvas

### ISSUE THREAD

**Issue by lmcnulty4** [d3/d3-geo#100]
Would it be possible to add an option allowing a callback to be registered and called on starting/ending the streaming of a geometry feature?

The idea would be to give users a bit more control (particularly when using a canvas) over the rendering of individual features.

I've tinkered myself with geoPath and added two new methods: _onFeatureBegin_ and _onFeatureEnd_. I've then passed these callbacks up to the actual streaming done within [stream.js](https://github.com/d3/d3-geo/blob/master/src/stream.js#L7) and called as so:

```js
var streamObjectType = {
  Feature: function(object, stream, onFeatureBegin, onFeatureEnd) {
    onFeatureBegin.call(object, object.properties);
    streamGeometry(object.geometry, stream);
    onFeatureEnd.call(object, object.properties);
  },
  FeatureCollection: function(object, stream, onFeatureBegin, onFeatureEnd) {
    var features = object.features, i = -1, n = features.length;
    while (++i < n) {
      onFeatureBegin.call(features[i], features[i].properties);
      streamGeometry(features[i].geometry, stream);
      onFeatureEnd.call(features[i], features[i].properties);
    }
  }
};
```

This lets me do things that I couldn't do even with a customised canvas context, such as:

```js
d3.geoPath()
  .context(context)
  .onStartFeature(e => {
    context.beginPath();
    context.fillStyle = ((e.id|0) % 2 === 1) ? "#FF0000" : "#000000"
  })
  .onEndFeature(e => {
    context.fill();
  });
```

which can be used as below:

```js
d3.json("https://unpkg.com/us-atlas@1/us/10m.json", function(error, us) {
  if (error) throw error;

  var features = topojson.feature(us, us.objects.states);
  features.features = features.features.map(e => {
      e.properties.id = e.id;
      return e;
  });

  path(features);
});
```
to render output looking like this: 

![image](https://cloud.githubusercontent.com/assets/12350206/25783042/4503e838-334d-11e7-9985-7dfb3c5e5ac7.png)

I know how to _almost_ achieve this using a custom canvas context, but if there are disjointed shapes (as, for example, in the case of Hawaii) I don't think it's possible to make it understand that these should be considered part of the same feature.

One of the uses of this would be to allow (with the help of custom context) keeping track of the rendered positions of features in real time and to know which rendered paths belong to which features (which islands are part of Hawaii, in the example above). I would like to use this in my [drag selector](https://github.com/lmcnulty4/d3dragSelector) plugin to allow it to work with canvases as well as SVG elements.

Is there a way to achieve this without adding functionality to d3, and if not, would it be possible to add this if you think it would be a useful feature?

**Comment by mbostock** [comment]
Hmm. The more consistent way of doing this (relative to the current API) would be to extend the *stream* interface to add *stream*.featureStart and *stream*.featureEnd methods. Then I would say something like this:

```js
var streamObjectType = {
  Feature: streamFeature,
  FeatureCollection: function(object, stream) {
    var features = object.features, i = -1, n = features.length;
    while (++i < n) streamFeature(features[i], stream);
  }
};

function streamFeature(object, stream) {
  stream.featureStart(object);
  streamGeometry(object.geometry, stream);
  stream.featureEnd();
}
```

But, I’m not sure I want to extend the stream interface to understand “features”—it’s intended only to represent geometry. I would probably implement your use case like this:

```js
topojson.feature(us, us.objects.states).features.forEach(e => {
  context.beginPath();
  context.fillStyle = ((e.id | 0) % 2 === 1) ? "#FF0000" : "#000000"
  path(e);
  context.fill();
});
```

If you feel like that’s not clean enough, you could define a standalone function to render a state:

```js
function renderState(e, context) {
  context.beginPath();
  context.fillStyle = ((e.id | 0) % 2 === 1) ? "#FF0000" : "#000000"
  path(e);
  context.fill();
}
```

Happy to keep discussing but I’m going to close this issue for now, and will reopen if we come to an agreement on API changes. Thank you!

**Comment by lmcnulty4** [comment]
Happy with closing this while we discuss!

To give a bit more background on why I'd like this, one of the things that is currently (I think, without API changes) impossible is having a data agnostic way of knowing what items (items which are rendered to a canvas) tie back to their original feature. To illustrate what I mean by this, I'll try to explain what I'm trying to do in a bit more detail. 

I'd like to extend a plugin I've written to allow it work with canvases. The plugin currently provides hit detection of SVG elements, allowing users to click and drag a rectangle to "select" elements, and it lets users configure what to do with these selected elements. In order to keep this hit detection as accurate and real-time as possible (to make it work even during transitions), the actual rendered positions of elements are used in determining whether or not they "hit". In this way it's agnostic about the data - to me and my users, it doesn't matter what data is being displayed or even what transitions or transformations that are applied, it only matters if the hit rectangle intersects them.

What I'd like to do with canvases is _something_ like:

```js
var path = d3.geoPath().context(context).projection(whatever);
var dragSelect = d3dragSelector().context(context).pathGenerator(path);
canvas.call(dragSelect.select());
```

Within my plugin I'd then replace the path's context with my own and keep track of all the graphics drawn to the context and use them in hit detection. But I'd need is some way to tie the data I've tracked back to individual features, in order to let users know which features were actually hit - closePath() isn't sufficient because of disjointed paths.

This way, it'd be data agnostic. It would let me not care about whatever data is being given to the geoPath generator, and let me determine what hits my selecting rectangle based purely on where they're rendered on the canvas.

It'd always be possible to do this if I interacted somehow with the actual data (like your examples above) but that feels far too messy for a small plugin to have to do, and is probably too onerous on users using the plugin to have to understand a more complicated API.

In terms of changing the current API, I think I'm not going to be very useful in discussing this at the moment. I'll have a deeper look and see if I can see a couple of candidate places for this change to fit in with as little footprint as possible.
